### PR TITLE
[BugFix] Fix bugs in GPU sampling and enable unit tests for dataloaders on the GPU

### PR DESCRIPTION
**Pull request by yaox12**
## Description

This fixes several bugs in the GPU sampling and thus enables unit tests `tests/pytorch/test_dataloader.py` for the GPU. 
`node_dataloader()/edge_dataloder()/neighbor_sampler_dataloder()` now pass the GPU unit tests.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the best of my knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change

## Changes
- Modify `python/dgl/dataloading/dataloader.py`: fix several bugs for GPU sampling
- Modify `src/graph/sampling/neighbor/neighbor.cc`: fix a corner case for GPU sampling
- Modify `tests/pytorch/test_dataloader.py`: enable unit tests for the GPU

### CODE REVIEW

**Comment by yaox12** [comment]
@BarclayII Please help review this, thanks.

**Review by BarclayII — APPROVED** [review]

**Comment by yaox12** [comment]
My fault, this requires the GPU version `edge_subgraph()` enabled in #3445. Let's get that merged first.

**Review by nv-dlasalle — APPROVED** [review]

File: tests/pytorch/test_dataloader.py [hunk 1]
```diff
@@ -218,8 +217,8 @@ def _check_device(data):
 def test_node_dataloader():
     sampler = dgl.dataloading.MultiLayerFullNeighborSampler(2)
 
-    g1 = dgl.graph(([0, 0, 0, 1, 1], [1, 2, 3, 3, 4]))
-    g1.ndata['feat'] = F.copy_to(F.randn((5, 8)), F.cpu())
+    g1 = dgl.graph(([0, 0, 0, 1, 1], [1, 2, 3, 3, 4])).to(F.ctx())
```
Comment (Rhett-Ying) [inline tests/pytorch/test_dataloader.py:220]: I think this change disable/skip the test of underlying graph copy from cpu to gpu within NodeDataLoader, as graph is not on the target device. After this change, graph is always on the target device already.

Comment (Rhett-Ying) [reply tests/pytorch/test_dataloader.py:220]: if graph is on cpu when created and target device is gpu, the sampling will fail without this change?

Comment (yaox12) [reply tests/pytorch/test_dataloader.py:220]: @Rhett-Ying Fixed in #3482
